### PR TITLE
Using Razor Components in Razor Pages and MVC projects

### LINKED ISSUE #13062: Shared Layout
From what I can tell, it is not possible for Blazor Pages (.razor) and Razor Pages (.cshtml) to share a single layout file.  Is that correct?  Whether it is or is not, I think it should be called out in this documentation, as it seems like a common case.  For instance, if you create a new Blazor application and include authorization, you are now in a hybrid environment where the Identity pages are all .cshtml, and have a separate layout from the Blazor pages.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: d64b8062-f910-a918-c5db-b2b67708602c
* Version Independent ID: 56c6cc6e-be94-d164-947c-cb14dbe91a19
* Content: [ASP.NET Core Blazor layouts](https://docs.microsoft.com/en-us/aspnet/core/blazor/layouts?view=aspnetcore-3.0#feedback)
* Content Source: [aspnetcore/blazor/layouts.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/layouts.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### LINKED ISSUE #13546: Using Razor Components in Razor Pages and MVC projects
The docs say that you can integrate server-side Razor Components with existing Razor Page project, but information is very limited and some elements are missing. It says that components must be wrapped in a div...

&lt;div id="counter"&gt;
    @(await Html.RenderComponentAsync&lt;Web.Components.Counter&gt;(new { }))
&lt;/div&gt;

...but this alone will not work. I found a reference on the page  https://docs.microsoft.com/en-us/aspnet/core/blazor/hosting-models?view=aspnetcore-3.0 that the following script tag needs to be added...

&lt;script src="_framework/blazor.server.js"&gt;&lt;/script&gt;

...which I had to place in my Pages/Shared/_Layout.cshtml and then I also had to add the following into my Startup.cs...

services.AddServerSideBlazor();
endpoints.MapBlazorHub();

This worked, but only for components on the index page, on all other routes and pages the HTML is rendered but the dynamic content does not. The missing part that I could not find documented anywhere is that you also need to place the following in the HEAD tag on your _Layout.cshtml

&lt;base href="~/" /&gt;

After this all server-side dynamic content can be rendered on any page or route of the project.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: be4c4d0c-149c-bde1-c7fa-f082916390ca
* Version Independent ID: 9d945996-ad02-bd4b-58ea-c465cc5cc74c
* Content: [Create and use ASP.NET Core Razor components](https://docs.microsoft.com/en-us/aspnet/core/blazor/components?view=aspnetcore-3.0#feedback)
* Content Source: [aspnetcore/blazor/components.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/components.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### LINKED ISSUE #16399: Your Partial Classes Documentation doesn't show how to use [Parameter] tag
I love that I can move code to partial class now, but I can't find any information about what namespace to add to make a property with a [Parameter] tag be available in the partial class.

It isn't obvious what namespace this is.

Thanks, 

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: be4c4d0c-149c-bde1-c7fa-f082916390ca
* Version Independent ID: 9d945996-ad02-bd4b-58ea-c465cc5cc74c
* Content: [Create and use ASP.NET Core Razor components](https://docs.microsoft.com/en-us/aspnet/core/blazor/components?view=aspnetcore-3.1)
* Content Source: [aspnetcore/blazor/components.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/components.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #13546
Fixes #13062
Fixes #16399
Addresses #16270
Addresses #16273

* Place the app-level concerns into the *Hosting Models* topic. Focus on the component inclusion in the *Components* topic. This results in less content in the uber components topic.
* Focuses on:
  * Component **_inclusion_** into pages and views.
  * Routable components for Razor Pages apps.
  * Routable components for MVC apps.
* This work also resolves the *how to use a shared layout* question from readers.
* Flesh out namespace issues early and cross-link to details further down in the topic.
* Add the `Microsoft.AspNetCore.Components` namespace to the partial class example. It's not strictly required in the current example, but as soon as something like a parameter is set up (`ParameterAttribute`), it's required.

Thanks @smadgerano! ⭐️
Thanks @DataJuggler! 🎸

### CODE REVIEW

File: aspnetcore/blazor/components.md [hunk 1]
```diff
@@ -46,15 +46,38 @@ Component members can be used as part of the component's rendering logic using C
...
 
-Components are ordinary C# classes and can be placed anywhere within a project. Components that produce webpages usually reside in the *Pages* folder. Non-page components are frequently placed in the *Shared* folder or a custom folder added to the project. To use a custom folder, add the custom folder's namespace to either the parent component or to the app's *_Imports.razor* file. For example, the following namespace makes components in a *Components* folder available when the app's root namespace is `WebApplication`:
+Components are ordinary C# classes and can be placed anywhere within a project. Components that produce webpages usually reside in the *Pages* folder. Non-page components are frequently placed in the *Shared* folder or a custom folder added to the project.
+
+Typically, a component's namespace is derived from the app's root namespace and the component's location (folder) within the app. If the app's root namespace is `BlazorApp` and the `Counter` component resides in the *Pages* folder:
+
+* The `Counter` component's namespace is `BlazorApp.Pages`.
+* The fully qualified component reference is `BlazorApp.Pages.Counter`.
```
Comment (danroth27) [inline aspnetcore/blazor/components.md]: I'm not sure what a "component reference" is...Do we define that somewhere? How about:

```suggestion
* The the fully qualified type name of the component is `BlazorApp.Pages.Counter`.
```

File: aspnetcore/blazor/components.md [hunk 2]
```diff
@@ -46,15 +46,38 @@ Component members can be used as part of the component's rendering logic using C
...
+* Public class members are referenced by the namespace and component class. If the `Counter` component has a component parameter named `MyParameter`, its member reference is `BlazorApp.Pages.Counter.MyParameter`.
```
Comment (danroth27) [inline aspnetcore/blazor/components.md]: I'm not sure what we mean by member references here. Components are just normal .NET classes with members. This sounds like we're inventing a new concept. Should this bullet just be removed?

File: aspnetcore/blazor/components.md [hunk 3]
```diff
@@ -46,15 +46,38 @@ Component members can be used as part of the component's rendering logic using C
...
 
 ```razor
-@using WebApplication.Components
+@using BlazorApp.Components
 ```
 
 ## Integrate components into Razor Pages and MVC apps
 
-Use components with existing Razor Pages and MVC apps. There's no need to rewrite existing pages or views to use Razor components. When the page or view is rendered, components are prerendered at the same time.
+Razor components can be integrated into Razor Pages and MVC apps. When the page or view is rendered, components are prerendered at the same time.
```
Comment (danroth27) [inline aspnetcore/blazor/components.md]: Technically components don't have to be prerendered with the page or view. You might just render a placeholder (`RenderMode.Server`). 

```suggestion
Razor components can be integrated into Razor Pages and MVC apps. When the page or view is rendered, components can be prerendered at the same time.
```

File: aspnetcore/blazor/components.md (hunk 2 above)
Comment (guardrex) [reply aspnetcore/blazor/components.md]: Yes ... I just made that up (based on it being a "member." I'll drop this bullet. It's obvious.

File: aspnetcore/blazor/hosting-models.md [hunk 4]
```diff
@@ -103,6 +103,187 @@ The graph is rerendered, and a UI *diff* (difference) is calculated. This diff i
...
 
+### Integrate Razor components into Razor Pages and MVC apps
+
+#### Use components in pages and views
+
+An existing Razor Pages or MVC app can integrate Razor components into pages and views:
+
+1. In the app's layout file (*_Layout.cshtml*):
+
+   * Add the following `<base>` tag to the `<head>` elements:
```
Comment (danroth27) [inline aspnetcore/blazor/hosting-models.md]:
```suggestion
   * Add the following `<base>` tag to the `<head>` element:
```

File: aspnetcore/blazor/hosting-models.md [hunk 5]
```diff
@@ -103,6 +103,187 @@ The graph is rerendered, and a UI *diff* (difference) is calculated. This diff i
...
+
+     ```html
+     <base href="~/" />
```
Comment (danroth27) [inline aspnetcore/blazor/hosting-models.md]: Your app might not be hosted at the root, right? I believe this needs to match whatever the base address of your app is.

**Comment by guardrex** [comment]
I put a line in with cross-link for *app base path* ...

> The `href` value (the *app base path*) in the preceding example assumes that the app's components are routable at the app's root path. For more information on setting the app base path, see \<xref:host-and-deploy/blazor/index#app-base-path>.

File: aspnetcore/blazor/hosting-models.md [hunk 6]
```diff
@@ -103,6 +103,189 @@ The graph is rerendered, and a UI *diff* (difference) is calculated. This diff i
...
+
+1. In the app's layout file (*_Layout.cshtml*):
+
+   * Add the following `<base>` tag to the `<head>` element:
+
+     ```html
+     <base href="~/" />
+     ```
+
+     The `href` value (the *app base path*) in the preceding example assumes that the app resides at the root URL path (`/`). For more information on setting the app base path, see <xref:host-and-deploy/blazor/index#app-base-path>.
```
Comment (danroth27) [inline aspnetcore/blazor/hosting-models.md]:
```suggestion
     The `href` value (the *app base path*) in the preceding example assumes that the app resides at the root URL path (`/`). If the app is a sub-application, follow the guidance in the *App base path* section of the <xref:host-and-deploy/blazor/index#app-base-path> article.
```

File: aspnetcore/blazor/hosting-models.md [hunk 7]
```diff
@@ -103,6 +103,189 @@ The graph is rerendered, and a UI *diff* (difference) is calculated. This diff i
...
+
+     The *_Layout.cshtml* file is located in the *Pages/Shared* folder in a Razor Pages app or *Views/Shared* folder in an MVC app.
+
+     If the app is a sub-application, follow the guidance in the *App base path* section of the <xref:host-and-deploy/blazor/index#app-base-path> article.
```
Comment (danroth27) [inline aspnetcore/blazor/hosting-models.md]: I suggest moving this above a bit where you're already talking about the base path. See my suggestion above.

**Pull request merged**